src/clj/fluree/db/flake/reasoner.cljc

```diff
@@ -235,21 +258,47 @@
             (recur r db**))
           db*)))))
 
+(defn find-duplicate-ids
+  [raw-rules]
+  (let [rule-ids (map first raw-rules)]
+    (->> rule-ids
+         frequencies
+         (filter #(< 1 (val %))))))
+
+(defn deduplicate-raw-rules
```

Comment:
I think this algorithm is needlessly wasteful. You go over the rules map multiple times to first find the frequencies, and then multiple `group-by` calls for each duplicate rule. I think it would be much more efficient to group the rules by the rule id once, and then filter the map returned by `group-by` that have more than one rule in the value of each rule entry.

Reply:
This function actually isn't getting used anymore, after talking with Brian we decided to instead go the route of overwriting duplicate ids and throwing a warning: https://github.com/fluree/db/pull/823#discussion_r1671003352. So I'll just get rid of this function.